packages/next-server/lib/router/sort-dynamic-routes.ts

```diff
@@ -0,0 +1,39 @@
+import { getRouteMatch } from "./utils";
+
+interface IDynamicRoute {
+  page: string
+  match: ReturnType<typeof getRouteMatch>
+}
+
+/*
+  Dynamic routes prioritization
+  1. Break up each path part '/a/$b' => [a, $b]
+  2. Prioritize path with non-dynamic parts over dynamic parts
+  3. Prioritize non-optional over optional '/a/$b' comes before '/a/$b$'
```

Comment:
I think you should add a test case for this specific assertion

Reply:
Yeah, we're still narrowing in on all of the cases we're looking at supporting

Reply:
Sure 👍 Anyway I don't think it's good to encourage people to put a non optional and an optional route in the same folder , but since it was here I thought a test case would be good 🙂